### DISCUSSION TITLE
Parsing Helm template syntax

### DISCUSSION THREAD

**Discussion by reiv** [eemeli/yaml#556] [Q&A] [answered]
I'm using this library in an (arguably misguided) attempt to parse and generate Helm templates, which use Go's `text/template` double curly brace syntax for interpolations. This works fine as long as everything is quoted or would otherwise be parsed as a string, like this:

```yaml
metadata:
  name: foo-{{ .Release.Name }}
```

but it quickly falls apart if you want to interpolate some non-string value like:

```yaml
spec:
  replicas: {{ .Values.replicaCount }}

# warn: Keys with collection values will be stringified due to JS Object restrictions: ...
```

Is it feasible to extend the parser to be able to deal with `{{ ... }}` outside of strings as a discrete node (similarly to how the library already lets you define custom scalars)?

**Comment by eemeli** [comment, marked as answer]
> Is it feasible to extend the parser to be able to deal with `{{ ... }}` outside of strings as a discrete node (similarly to how the library already lets you define custom scalars)?

No, because that's valid YAML that has a non-scalar meaning, and it's parsed as an object with an object key already by the lexer.

However, you could add a wrapper around the `yaml` parser to make it work:
1. Replace all of the `{{ ... }}` sequences with some string that's a valid YAML plain scalar, like `__TEMPLATE_VAR_#__` where `#` is an identifier that you keep track of.
2. Parse the result with `parseDocument()`.
3. Use a [visitor](https://eemeli.org/yaml/#finding-and-modifying-nodes) to find and replace the `__TEMPLATE_VAR_#__` strings with their original values.

**Reply by reiv** [reply]
Thank you for the suggestion. The preprocessor approach seems like a reasonable way of doing it (and also rather apt since Helm is a preprocessor itself).
